lib/restforce/db/cleaner.rb

```diff
@@ -62,7 +62,45 @@ def invalid_salesforce_ids
         valid_ids = valid_salesforce_ids
         all_ids = all_salesforce_ids
 
-        all_ids - valid_ids
+        invalid_ids = all_ids - valid_ids
+        confirmed_invalid_salesforce_ids(invalid_ids)
+      end
+
+      # In order to ensure that we don't generate any SOQL queries which are too
+      # long to send across the wire to Salesforce, we need to batch IDs for our
+      # queries. For a conservative cap of 8,000 characters per GET query, at 27
+      # encoded characters per supplied ID (18 characters and 3 three-character
+      # entities), 250 IDs gives us a buffer of around 1250 spare characters to
+      # work with for the rest of the URL and query string.
+      #
+      # In practice, there should rarely/never be this many invalidated records
+      # at once during a single worker run.
+      MAXIMUM_IDS_PER_QUERY = 250
+
+      # Internal: Get the IDs of records which have been proposed as invalid and
+      # do not in fact appear in response to a time-insensitive query with the
+      # requisite conditions applied.
+      #
+      # NOTE: This double-check step is necessary to prevent an inaccurate
+      # Salesforce server clock from sending records back in time and forcing
+      # them to show up in a query running after-the-fact.
+      #
+      # proposed_invalid_ids - An Array of String Salesforce IDs to test against
+      #                        the Salesforce server.
+      #
+      # Returns an Array of IDs.
+      def confirmed_invalid_salesforce_ids(proposed_invalid_ids)
+        proposed_invalid_ids.each_slice(MAXIMUM_IDS_PER_QUERY).inject([]) do |invalid_ids, ids|
+          # Get a subset of the proposed list of IDs that corresponds to
+          # records which are still valid for any parallel mapping.
+          valid_ids = parallel_mappings.flat_map do |mapping|
+            mapping.salesforce_record_type.all(
+              conditions: "Id in ('#{ids.join("','")}')",
+            ).map(&:id)
+          end
+
+          invalid_ids + (ids - valid_ids)
```

Comment:
I thought that we were going to also log any occurrences in which we trip this condition.